cylc/flow/graph_parser.py

```diff
@@ -922,46 +924,46 @@ def _compute_triggers(
+                    if not expr:
+                        # infer :succceed-all for lone family node
+                        output = QUAL_FAM_SUCCEED_ALL
                 elif output.startswith("finish"):
                     if optional:
                         raise GraphParseError(
                             f"Family pseudo-output {name}:{output} can't be"
                             " optional")
                     # But implicit optional for the real succeed/fail outputs.
                     optional = True
-                try:
-                    outputs = self.__class__.fam_to_mem_output_map[output]
-                except KeyError:
-                    # Illegal family trigger on RHS of a pair.
-                    raise GraphParseError(
-                        f"Illegal family trigger: {name}:{output}"
-                    ) from None
+                if output:
+                    try:
+                        outputs = self.__class__.fam_to_mem_output_map[output]
+                    except KeyError:
+                        # Illegal family trigger on RHS of a pair.
+                        raise GraphParseError(
+                            f"Illegal family trigger: {name}:{output}"
+                        ) from None
+                else:
+                    outputs = ['']
             else:
                 fam = False
-                # RHS: don't infer ":succceeded", except:
-                if (
-                    # infer ":succeeded? from explicit "?"
-                    (not output and optional) or
-                    # and for "None => right" which is really a lone node.
-                    (not output and not expr)  # lone node: "None => right"
-                ):
-                    output = TASK_OUTPUT_SUCCEEDED
-                else:
+                if output:
                     # Convert to standard output names if necessary.
                     output = TaskTrigger.standardise_name(output)
+                else:
+                    # RHS
+                    if optional or not expr:
+                        # infer ":succeeded? from explicit "?"
+                        # infer ":succeeded" from lone node (None => lone)
+                        output = TASK_OUTPUT_SUCCEEDED
                 rhs_members = [name]
-                outputs = [output]
+                outputs = [output]  # can be [None]
 
             for mem in rhs_members:
                 if not offset:
                     # Nodes with offsets on the RHS do not define triggers.
                     self._set_triggers(mem, suicide, trigs, expr, orig_expr)
                 for output in outputs:
-                    if output is None:
-                        # no explicit output on RHS
-                        continue
-                    # But they must be consistent with output optionality.
-                    self._set_output_opt(mem, output, optional, suicide, fam)
+                    if output:
```

Comment:
More elegant and less nested as a comprehension IMO

```python
[
    self._set_output_opt(mem, output, optional, suicide, fam)
    for output in outputs if output
]

```

Reply:
Well it's elegant, but it unnecessarily constructs a list.

Reply:
```python
(
    self._set_output_opt(mem, output, optional, suicide, fam)
    for output in outputs if output
)

Reply:
That constructs a generator object.  All we  want to do here is call a function for each output, for the function side effects, so I think a loop is fine for that

Reply:
Could construct a list of non-null outputs first, in a comprehension, to avoid the `if` inside the loop, but is that worth it?